Add support for setting message retention policy for individual streams.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This adds support for setting message retention policy for individual streams.
Only admins are allowed to change it or set it while creating streams.



**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
While creating stream
![Screenshot (127)](https://user-images.githubusercontent.com/35494118/84709088-46b20280-af7f-11ea-9da5-57486c95e54f.png)

While updating stream permissions
![Screenshot (128)](https://user-images.githubusercontent.com/35494118/84709103-4ade2000-af7f-11ea-8b6c-50357246fbbd.png)




<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -3610,6 +3610,20 @@ def do_change_stream_description(stream: Stream, new_description: str) -> None:
+        name=stream.name,
+    )
+    send_event(stream.realm, event, can_access_stream_user_ids(stream))
+
```
I am not sure whether I need to mention this change of a new property for stream update event in changelog.md. 
Let me know if this is required.

timabbott commented:
@pragatiagrawal31 @mateuszmandera can you help review this?

pragatiagrawal31 commented on static/js/stream_edit.js:
```diff
@@ -63,6 +63,51 @@ exports.get_users_from_subscribers = function (subscribers) {
+    if (page_params.realm_message_retention_days === -1) {
+        return i18n.t("(forever)");
+    }
+    return i18n.t("(__message_retention_days__ days)", {message_retention_days: page_params.realm_message_retention_days});
```
`__message_retention_days__` will be rendered as empty when `page_params.realm_message_retention_days` is `null`.

pragatiagrawal31 replied:
In my PR, https://github.com/zulip/zulip/pull/14938, we have a pending discussion of what should be the best approach to set the `message_retention_days` to "Forever".

In master, both `null`, as well as `-1`, means "Forever", but this is not coherent with Streams case where `-1` means "disable retention policy for this stream unconditionally", so setting `-1` at the realm level implies "No matter what we want the message to retain forever even if some streams have different value".

sahil839 replied:
Message retention docs [here](https://zulipchat.com/help/message-retention-policy) says 
`Setting a retention policy for individual streams, to just delete messages on those specific streams.` and it suggests that we will delete the messages on the streams even if realm setting is set to forever.

pragatiagrawal31 replied:
... which is bit confusing.

I see how you managed the "Forever" case for the stream by passing a string "forever" rather than sticking to a number, @timabbott @mateuszmandera what are your opinions on this?

This makes me wonder, at the realm level rather than "Retain forever" it should mean "Retain forever unless specified at stream level"?

sahil839 replied:
I also have gone through some code in `zerver/lib/retention.py` and we do not handle case with realm value set to -1. It seems like in this file we assume that null means forever.

pragatiagrawal31 replied:
> Message retention docs [here](https://zulipchat.com/help/message-retention-policy) says
> `Setting a retention policy for individual streams, to just delete messages on those specific streams.` and it suggests that we will delete the messages on the streams even if realm setting is set to forever.

(GitHub real-time page update is kinda broken, that's why I didn't notice you reply)

yes, that's true, I was saying from a developers perspective and maybe from API standpoint too, that `-1` at `Stream` side means something else. Related discussion: https://github.com/zulip/zulip/pull/14903#issuecomment-626204683

sahil839 replied:
Yeah -1 for stream means we will retain forever inspite of realm-setting value being a finite postive number.
I was thinking the same as you mentioned in that comment, that we could use None as retain forever rather than using -1. Maybe we can wait on decision of whether to use -1 or None for the forever case in realm-level settings

mateuszmandera replied:
The confusion is due to the fact that `Realm.message_retention_days` does not mean a strict "keep messages forever".

`Realm.message_retention_days` being NULL means the natural thing "no realm-wide policy set", which is the default situation and we're not archiving old messages. So then, if a stream sets it retention policy then we're going to archive the stream's message according to that - there's no contradiction with the realm's policy when the realm has no policy.

Streams have the special option of declaring "keep my messages forever" (through the value `-1`), because that's something potentially useful if the admin wants to exclude certain streams from the realm-wide policy.  The equivalent doesn't make sense for an entire realm, a realm can merely have no policy declared (`NULL`)

Does that help clear things up? :thinking:

sahil839 replied:
Thanks, it does makes a lot of things clear to me. 
So, we are changing realm-wide policy to be NULL for forever case and are removing use of -1 as implemented in #14938, correct?

mateuszmandera replied:
So far we're not changing the meaning of any values for the `message_retention_days` field on the backend models. The problem in the PR is "how to convey `no policy set` to the server through the API" - because the webapp can't send the `null` value, because that already has a meaning in the setting system and means "no change to this setting has been made". So for now in the PR's code `-1` is used to represent that. But that gets very confusing once streams come into play as you can see...

mateuszmandera replied:
Okay I just got caught up on https://chat.zulip.org/#narrow/stream/2-general/topic/message.20retention.20for.20streams/near/903209 So if @sahil839 is going with string values, so should #14938, instead of sending the `-1`, this is much better.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2257,6 +2257,18 @@ paths:
+                              and a null value means that the stream will follow organization
+                              level setting.
+
+                              **Changes**: New in Zulip 2.2 (feature level 16)
```
Use a period at the end of these sentences.

timabbott commented on static/js/stream_ui_updates.js:
```diff
@@ -143,7 +143,11 @@ exports.update_stream_privacy_type_icon = function (sub) {
 exports.update_stream_subscription_type_text = function (sub) {
     const stream_settings = stream_edit.settings_for_sub(sub);
-    const html = render_subscription_type(sub);
+    const template_data = sub;
```
You should do make a new object based on a copy of sub to do this sort of template work.  The `sub` objects returned by `settings_for_sub` are singletons, so mutating them is a bad idea.

timabbott commented on zerver/views/streams.py:
```diff
@@ -125,6 +127,15 @@ def check_if_removing_someone_else(user_profile: UserProfile,
+    if value == "realm_default":
+        return None
+    if isinstance(value, str) or value < 0:
+        raise JsonableError(_("Invalid value of message_retention_days"))
```
We should use the standard exception from `zerver/lib/request.py`; I'll push back with this change (you'll need to fix the tests).

timabbott commented on zerver/views/streams.py:
```diff
@@ -224,10 +235,15 @@ def update_stream_backend(
     stream = access_stream_for_delete_or_update(user_profile, stream_id)
+    if message_retention_days is not None:
+        message_retention_days_value = parse_message_retention_days(message_retention_days)
+        do_change_stream_message_retention_days(stream, message_retention_days_value)
```
We should enforce the user is a realm owner here (and raise the right exception!), not just an admin, right?

timabbott replied:
Maybe worth doing in `parse_message_retention_days`, and having that take the UserProfile as an argument.

We also need to check `plan_type`; @hackerkid can help with that if you're not sure how.

sahil839 replied:
During CZO discussion, you told that we will allow admins to set this property.
If we want to change, we can do owners.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -63,6 +63,51 @@ exports.get_users_from_subscribers = function (subscribers) {
+
+    if (message_retention_days === null)  {
+        message_retention_days = page_params.realm_message_retention_days;
+    }
```
I think you can write this more cleanly by having this be the first thing you do.  then that first if statement doesn't need to have so much complexity.

sahil839 replied:
@timabbott  I am not sure that it will become clean because, in that case also we want to check for if stream value is set to -1 and in this case also both organization-level and stream-level setting will be forever and we do not want to show any text.

I also have missed the case of `page_params.realm_message_retention_days` being null here. I was thinking of something like this for the case of not showing any text:
`[-1, null].includes(page_params.realm_message_retention_days) && [-1, null].includes(sub.message_retention_days)`
I have seen this being used in other places, is it fine to use this?

timabbott commented on static/js/stream_create.js:
```diff
@@ -169,6 +169,13 @@ function create_stream() {
 
```
generate_secrets.py.isorted should not be here.

sahil839 replied:
I do not understand this comment. Maybe this is a typo.

timabbott replied:
Oh, I think I removed it in my push back yesterday, but you had added an extra file by that name  to the codebase.

timabbott commented on static/js/stream_create.js:
```diff
@@ -169,6 +169,13 @@ function create_stream() {
 
     data.stream_post_policy = JSON.stringify(stream_post_policy);
 
+    let message_retention_days = $('#stream_creation_form select[name=stream_message_retention_setting]').val();
```
Let's use a better variable name for this temporary value; `message_retention_selection`, perhaps?

timabbott commented:
Thanks for working on this @sahil839!

I merged 8bd1429f4c396f4edbd77f6bf46aa0f77fae4ff0 and then pushed back here after making a few textual edits to the first commit.  Let me know when you have a revised version ready.

sahil839 commented:
It would be nice if we can do a final decision on #14938. If it is decided that we are using NULL in database for forever case of realm-level-setting instead of -1, then I can change accordingly.

timabbott commented:
I don't think we can use `null` for that purpose at the stream level, since `null` means inherit-from-realm.  So we're going to stick with `-1` in the database for forever in any case.

sahil839 commented:
I am saying null for realm-level forever settings as implemented in #14938.

timabbott commented:
I'd prefer keeping -1 for that.  It's more consistent.

-Tim Abbott (mobile)

On Thu, Jun 18, 2020, 01:57 Sahil Batra <notifications@github.com> wrote:

> I am saying null for realm-level forever settings as implemented in #14938
> <https://github.com/zulip/zulip/pull/14938>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/15395#issuecomment-645881180>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAU6NWWTTRUKSBRPOQZM6HDRXHJGXANCNFSM4N62ZKJQ>
> .
>

sahil839 commented:
So, as we are showing realm level setting in the drop-down with the 'follow organization level settings option', do we want to show different text for cases of realm level settings being -1 and `null` or we can show forever as realm level setting being null(not set)  essentially means messages are retained forever.

sahil839 commented:
@timabbott I have updated the PR.
- I have added the case for realm-level settings being null.
- Have added restriction for allowing only owners and also checks for plan type.
- I have not yet changed the implementation as asked [here](https://github.com/zulip/zulip/pull/15395#discussion_r442053443) and have asked a query for this below that comment. I am not sire but I think we can do it more cleanly with
`[-1, null].includes(sub.message_retention_days) &&[-1,null].includes(page_params.realm_message_retention_days)`. This style has been used in other places in code too.

timabbott commented:
Merged as the series ending with f3767b909ce057de90eeb7f0055a51aa24e597ae, after just comment/documentation tweaks, thanks @sahil839!

I'm not sure the `.includes` lines would really be any better, so I just added a comment explaining the purpose of the logic.
